Mathlib/Tactic/Order.lean

```diff
@@ -248,27 +227,47 @@ def orderCore (only? : Bool) (hyps : Array Expr) (negGoal : Expr) (g : MVarId) :
     let TypeToAtoms ← collectFacts only? hyps negGoal
     for (type, (idxToAtom, facts)) in TypeToAtoms do
       let some orderType ← findBestOrderInstance type | continue
-      let facts : Array AtomicFact ← match orderType with
-      | .pre => preprocessFactsPreorder facts
-      | .part => preprocessFactsPartial facts idxToAtom
-      | .lin => preprocessFactsLinear facts idxToAtom
       trace[order] "Working on type {← ppExpr type} ({orderType})"
       let atomsMsg := String.intercalate "\n" <| Array.toList <|
         ← idxToAtom.toArray.sortDedup.mapM
           fun ⟨idx, atom⟩ => do return s!"#{idx} := {← ppExpr atom}"
       trace[order] "Collected atoms:\n{atomsMsg}"
       let factsMsg := String.intercalate "\n" (facts.map toString).toList
       trace[order] "Collected facts:\n{factsMsg}"
-      let mut graph ← Graph.constructLeGraph idxToAtom.size facts idxToAtom
-      graph ← updateGraphWithNltInfSup graph idxToAtom facts
+      let facts ← replaceBotTop facts idxToAtom
+      let processedFacts : Array AtomicFact ← preprocessFacts facts idxToAtom orderType
+      let factsMsg := String.intercalate "\n" (processedFacts.map toString).toList
+      trace[order] "Processed facts:\n{factsMsg}"
+      let mut graph ← Graph.constructLeGraph idxToAtom.size processedFacts
+      graph ← updateGraphWithNltInfSup graph idxToAtom processedFacts
       if orderType == .pre then
-        let some pf ← findContradictionWithNle graph idxToAtom facts | continue
+        let some pf ← findContradictionWithNle graph idxToAtom processedFacts | continue
         g.assign pf
         return
-      else
-        let some pf ← findContradictionWithNe graph idxToAtom facts | continue
+      if let some pf ← findContradictionWithNe graph idxToAtom processedFacts then
         g.assign pf
         return
+      -- if fast procedure failed and order is linear, we try `omega`
+      if orderType == .lin then
+        let .succ u ← getLevel type | {
+          trace[order] "Translation to `Int` failed: unexpected Prop instead of type.";
+          throwError "Unexpected Prop" }
+        let type : Q(Type u) := type
+        let instLinearOrder ← synthInstanceQ q(LinearOrder $type)
+        let (_, factsNat) ← translateToInt type instLinearOrder idxToAtom facts
```

Comment:
```suggestion
        -- Here we only need to translate the hypotheses, since the goal will remain to derive `False`.
        let (_, factsNat) ← translateToInt type instLinearOrder idxToAtom facts
```